### PR TITLE
Add coverage for cuda tests

### PR DESCRIPTION
**Pull request by EmilyBourne**
Move cuda tests to run with linux tests. Once they have run the coverage results from both are combined to test the coverage. This PR also includes the bug fixes for the coverage checker scripts

### CODE REVIEW

File: .github/actions/coverage_collection/action.yml [hunk 1]
```diff
@@ -6,7 +6,6 @@ runs:
     - name: Coverage collection
       run: |
         coverage combine
-        coverage xml
```
Comment (bauom) [inline .github/actions/coverage_collection/action.yml:9]: why did we remove this but we still used it in the Cuda coverage collection

File: .github/workflows/Github_pytest.yml [hunk 2]
```diff
@@ -95,10 +89,94 @@ jobs:
...
+        with:
+          name: cuda-coverage-artifact
+          path: .coverage
+          retention-days: 1
+
+  CoverageCollection:
+
+    runs-on: ubuntu-latest
+    needs: [Linux, Cuda]
+    if: ${{ always() && needs.Linux.result == 'success' && needs.Cuda.result != 'failure' }}
```
Comment (bauom) [inline .github/workflows/Github_pytest.yml:130]: is there more that `success`, `failure` in the result ?

**Review by bauom — CHANGES_REQUESTED** [review]

File: .github/actions/coverage_collection/action.yml (hunk 1 above)
Comment (EmilyBourne) [reply .github/actions/coverage_collection/action.yml:9]: `coverage combine` combines `.coverage.X` files into a `.coverage` file. Once we have a `.coverage` file describing all the tests, we can call `coverage xml` which converts a `.coverage` file into an XML summary.

If I leave this here then the output is the XML summary not the `.coverage` file and then it is much harder to combine the outputs from the linux tests and the cuda tests

File: .github/workflows/Github_pytest.yml (hunk 2 above)
Comment (EmilyBourne) [reply .github/workflows/Github_pytest.yml:130]: It can also be `skipped`

File: .github/actions/coverage_collection/action.yml (hunk 1 above)
Comment (bauom) [reply .github/actions/coverage_collection/action.yml:9]: Oh I see thank you.

**Review by bauom — APPROVED** [review]

**Pull request merged**